notebooks/rocoto.ipynb

```diff
@@ -0,0 +1,1013 @@
+    "<!--cell 10-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "56d04e6b-5fa2-4c93-ac2f-c95fa23c888e",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "workflow:\n",
+      "  attrs:\n",
+      "    scheduler: slurm\n",
+      "  cycledef:\n",
+      "    - spec: 202410290000 202410300000 06:00:00\n",
+      "  tasks:\n",
+      "    task_greet:\n",
+      "      cores: 1\n",
+      "      walltime: 00:00:10\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%bash\n",
+    "cat fixtures/rocoto/err-workflow.yaml"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f102c6ca-c8a5-4f9c-8e40-b5fa040bbeab",
+   "metadata": {},
+   "source": [
+    "When any type of validation error occurs, `realize()` will raise an exception indicating what type of error occurred. The number of validation errors present and their locations within the workflow structure are given.\n",
+    "<!--cell 12-->"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "16704f19-cbca-4765-8c72-16512fc96e9b",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "[2024-10-28T12:34:55]    ERROR 3 UW schema-validation errors found\n",
+      "[2024-10-28T12:34:55]    ERROR Error at workflow -> attrs:\n",
+      "[2024-10-28T12:34:55]    ERROR   'realtime' is a required property\n",
+      "[2024-10-28T12:34:55]    ERROR Error at workflow -> tasks -> task_greet:\n",
+      "[2024-10-28T12:34:55]    ERROR   'command' is a required property\n",
+      "[2024-10-28T12:34:55]    ERROR Error at workflow:\n",
+      "[2024-10-28T12:34:55]    ERROR   'log' is a required property\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "YAML validation errors\n"
+     ]
+    }
+   ],
+   "source": [
+    "try:\n",
+    "    rocoto.realize(\n",
+    "        config=Path('fixtures/rocoto/err-workflow.yaml'),\n",
+    "        output_file='tmp/err-workflow.xml'\n",
+    "    )\n",
+    "except Exception as e:\n",
+    "    print(e)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cbad7aef-02b6-43a3-9241-349affa2f71c",
+   "metadata": {},
+   "source": [
+    "### Entities and Cyclestrings\n",
+    "\n",
+    "Constants called entities may be defined so that their values can be referenced throughout the rest of the workflow. These are defined in an `entities:` block, with their names and values given as keys and values in the YAML. Below an entity named `LOG` is defined with a string value. This value is referred to in the other parts of the workflow with the syntax `&ENTITY_NAME;`. In this case, note the `&LOG;` entity within the `log:` block.\n",
```

Comment:
```suggestion
    "Constants called entities may be defined so that their values can be referenced throughout the rest of the workflow. These are defined in an `entities:` block, with their names and values given as keys and values in the YAML. Below, an entity named `LOG` is defined with a string value. This value is referred to elsewhere in the workflow with the syntax `&ENTITY_NAME;`. In this case, note the `&LOG;` entity within the `log:` block.\n",
```